### PR TITLE
Move program-2022 tests to new crate

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
A circular dependency with spl-token-client prevents publishing spl-token-2022 to crates.io

#### Solution
- Move tests that need spl-token-client to their own crate.
- Remove various features no longer needed.

Open to bikeshedding on what the test crate should be called; I just followed a precedent from the monorepo

### CODE REVIEW

File: associated-token-account/program/Cargo.toml [hunk 1]
```diff
@@ -10,15 +10,11 @@ edition = "2018"
 [features]
 no-entrypoint = []
 test-bpf = []
-# When spl-token depends on spl-token-client, which depends on this with no-entrypoint,
-# all tests fail! Normally, we want no-entrypoint, except when testing spl-token
-no-token-entrypoint = [ "spl-token/no-entrypoint" ]
-default = [ "no-token-entrypoint" ]
```
Comment (joncinque) [inline associated-token-account/program/Cargo.toml:16]: :heart:

File: token/rust/Cargo.toml [hunk 2]
```diff
@@ -7,18 +7,13 @@ name = "spl-token-client"
 repository = "https://github.com/solana-labs/solana-program-library"
 version = "0.0.1"
 
-# When spl-token-2022 depends on this with no-entrypoint, all tests fail!
-# Normally, we want no-entrypoint, except when testing spl-token-2022
-[features]
-default = [ "spl-token-2022/no-entrypoint", "spl-associated-token-account/no-token-entrypoint" ]
-
```
Comment (joncinque) [inline token/rust/Cargo.toml:14]: :heart:

**Review by joncinque — APPROVED** [review]
Looks great! Makes me a little embarrassed that I didn't do this from the start

**Comment by mvines** [comment]
> > Makes me a little embarrassed that I didn't do this from the start
> 
> Ditto, wasn't thinking ahead to publishing.

I briefly looked into adapting https://github.com/solana-labs/solana/blob/master/ci/order-crates-for-publishing.py#L4 for the SPL repo, which would have flagged the circular dep right from the start.  The non-uniform crate names in the SPL repo gave me a bit of headache and I gave up for now.  But if this happens again somewhere in SPL then it's prolly worth bringing in that CI extra check

**Review by mvines — APPROVED** [review]
thanks for doing this

**Pull request merged**